deeptrack/models/gans/pcgan.py

```diff
@@ -0,0 +1,152 @@
+import tensorflow as tf
+from tensorflow.python.keras.utils.generic_utils import default
+from .utils import as_KerasModel
+
+layers = tf.keras.layers
+
+
+@as_KerasModel
+class PCGAN(tf.keras.Model):
+    """Creates and compiles a conditional generative adversarial
+       neural network (cgan) with an additional perceptual discriminator. 
+    """
+
+    def __init__(
+        self,
+        generator=None,
+        discriminator=None,
+        discriminator_loss=None,
+        discriminator_optimizer=None,
+        discriminator_metrics=None,
+        assemble_loss=None,
+        assemble_optimizer=None,
+        assemble_loss_weights=None,
+        content_discriminator="DenseNet121",
+        metrics=[],
+        **kwargs
+    ):
+        super(PCGAN).__init__()
+
+        # Build and compile the discriminator
+        self.discriminator = discriminator
+        self.discriminator.compile(
+            loss=discriminator_loss,
+            optimizer=discriminator_optimizer,
+            metrics=discriminator_metrics,
+        )
+
+        # perceptual discriminator (pre-trained)
+        self.content_discriminator = tf.keras.Sequential(
+            [
+                layers.Lambda(
+                    lambda img: layers.Concatenate(axis=-1)([img] * 3)
+                ),
+                getattr(tf.keras.applications, content_discriminator)(
+                    include_top=False,
```

Comment:
Would this crash if you changed content_discriminator to something else? What are the available networks? Could we make it so one could train their own network and pass it?

Reply:
The previous implementation only supported using models available in [Keras Aplications](https://www.tensorflow.org/api_docs/python/tf/keras/applications). Now, it is also possible to provide a path to model weights other than ImageNet (e.g., custom training weights). Furthermore, the user can pre-train his perceptual model separately and provide it directly to the network, just as it would be done with the generator and discriminator networks.

Reply:
Please see the review below for further details!